lib/pbench/server/api/resources/upload_api.py

```diff
@@ -56,303 +147,292 @@ def __init__(self, config, logger):
                 __name__, "pbench-server", "rest_max_content_length", self.logger
             )
         )
+        self.temporary = config.ARCHIVE / FileTree.TEMPORARY
+        self.temporary.mkdir(mode=0o755, parents=True, exist_ok=True)
+        self.logger.info("Configured PUT temporary directory as {}", self.temporary)
 
     @Auth.token_auth.login_required()
     def put(self, filename: str):
+
+        # Used to record what steps have been completed during the upload, and
+        # need to be undone on failure
+        recovery = Cleanup(self.logger)
+
         try:
-            username = Auth.token_auth.current_user().username
-        except Exception as exc:
-            self.logger.error("Error verifying the username: '{}'", exc)
-            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+            try:
+                username = Auth.token_auth.current_user().username
+            except Exception:
+                username = None
+                raise CleanupTime(
+                    HTTPStatus.INTERNAL_SERVER_ERROR, "Error verifying username"
+                )
 
-        if os.path.basename(filename) != filename:
-            msg = "File must not contain a path"
-            self.logger.warning(
-                "{} for user = {}, file = {!a}", msg, username, filename,
-            )
-            abort(HTTPStatus.BAD_REQUEST, message=msg)
+            controller = request.headers.get("controller")
```

Comment:
We do record the controller, but it's essentially becoming metadata; and since I'm already opening the `metadata.log` of the incoming tarball, we could get it from there instead of requiring it as a header.

I considered doing that in #2575 but decided it was something we should consider separately. I'm thinking of pulling everything in `metadata.log` into Dataset `Metadata` to make it easily available, and this change might well make sense as part of that.